server/broadcast.go

```diff
@@ -100,6 +123,23 @@ func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
 		   To avoid a runtime search of the session list under lock, simply
 		   fixup the session list at selection time by retrying the selection.
 		*/
+		if _, ok := bsm.sessMap[sess.OrchestratorInfo.Transcoder]; ok {
+			if monitor.Enabled && bsm.lastSess != nil && bsm.lastSess.OrchestratorInfo.Transcoder != sess.OrchestratorInfo.Transcoder {
+				monitor.OrchestratorSwapped()
+			}
+			bsm.lastSess = sess
+			return sess
+		}
+
+		// Last session got removed from map (possibly due to a failure) so stop tracking its in-flight segments
+		if bsm.lastSess != nil && sess.OrchestratorInfo.Transcoder == bsm.lastSess.OrchestratorInfo.Transcoder {
+			bsm.lastSess.SegsInFlight = nil
```

Comment:
If the last session was removed from the map should we also set `bsm.lastSess = nil` here since we know that we wouldn't want to use the last session in the next iteration of the loop?

Reply:
This wasn't a hard requirement as we would also test for `len(bsm.lastSess.SegsInFlight) > 0` in next iteration and then end up using a new session or nil.
But adding this definitely makes the code more explicit/clear so done that in the update!